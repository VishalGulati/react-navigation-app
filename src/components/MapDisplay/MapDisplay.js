import React, { Component } from 'react';
import './MapDisplay.css';
import PropTypes from 'prop-types';
import { getLocationOnMap } from '../../config/utilities';

/**
 * The Right halve of main page that displays the map with route based on search result.
 */
class MapDisplay extends Component {
  /**
 * @name showDefaultPosition
 * @description This method shows default position on map when the app load
 */
  showDefaultPosition = () => {
    this.directionsService = new window.google.maps.DirectionsService();
    this.directionsDisplay = new window.google.maps.DirectionsRenderer();
    this.map = new window.google.maps.Map(this.mapWrapper, {
      center: {
        lat: 51.509865,
        lng: -0.118092
      },
      zoom: 12,
      mapTypeId: 'roadmap'
    });
  };

  /**
 * @name diaplayRoute
 * @description This method displays the route on map once search request is successful
 */
  diaplayRoute = () => {
    this.directionsDisplay.setMap(this.map);
    const middlePoints = [...this.props.route];
    const origin = middlePoints.shift();
    const destination = middlePoints.pop();
    const wayPointsRoute = middlePoints.map(middlePt => {
      return {
        location: getLocationOnMap(middlePt[0], middlePt[1]),
        stopover: true
      };
    });
    this.directionsService.route(
      {
        origin: getLocationOnMap(origin[0], origin[1]),
        destination: getLocationOnMap(destination[0], destination[1]),
        waypoints: wayPointsRoute,
        optimizeWaypoints: true,
        travelMode: 'DRIVING'
      },
      (response, status) => {
        if (status === 'OK') {
          this.directionsDisplay.setDirections(response);
        } else {
          window.alert('Directions request failed with status: ' + status);
        }
      }
    );
  };

  /**
 * @name resetMap
 * @description This method removes the previously displayed route from map
 */
  resetMap = () => {
    this.directionsDisplay.setMap(null);
  };

  /**
 * @name componentDidUpdate
 * @description React component lifecycle hook
 */
  componentDidUpdate(prevProps) {
    if (this.props.showRoute) {
      this.diaplayRoute();
    } else {
      this.resetMap();
    }
  }

  componentDidMount() {
    this.showDefaultPosition();
  }

  render() {
    return (
      <div className="col-xs-12 col-md-8">
        <div id="googleMap" ref={el => (this.mapWrapper = el)} />
      </div>
    );
  }
}

MapDisplay.propTypes = {
  showRoute: PropTypes.bool,
  route: PropTypes.array
};

export default MapDisplay;
