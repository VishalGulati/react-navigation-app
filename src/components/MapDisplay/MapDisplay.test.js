import React from 'react';
import ReactDOM from 'react-dom';
import MapDisplay from './MapDisplay';
import { shallow } from 'enzyme';

const wrapper = shallow(<MapDisplay />);

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<MapDisplay />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('Component has container div to hold the google map', () => {
  expect(wrapper.find('div#googleMap').length).toEqual(1);
});

it('snapshot testing', () => {
  expect(wrapper).toMatchSnapshot();
});
