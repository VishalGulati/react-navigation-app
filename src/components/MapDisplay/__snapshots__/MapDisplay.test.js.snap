// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MapDisplay />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        id="googleMap"
      />,
      "className": "col-xs-12 col-md-8",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "id": "googleMap",
      },
      "ref": [Function],
      "rendered": null,
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          id="googleMap"
        />,
        "className": "col-xs-12 col-md-8",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "id": "googleMap",
        },
        "ref": [Function],
        "rendered": null,
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
